### DISCUSSION TITLE
Redundant resolving of packages

### DISCUSSION THREAD

**Discussion by yha** [fonsp/Pluto.jl#1955] [Questions about using Pluto]
Sometimes Pluto re-resolves packages and says "Restart notebook (recommended)" when I just moved a `using` to a different place.

![recording](https://user-images.githubusercontent.com/4170948/155620907-3793fedc-412d-475d-bac9-e2c0d7dc6a67.gif)

**Comment by icweaver** [comment]
I think the problem is that you are not just moving an already imported package to a different place. When you hit backspace on the second cell to delete it, you are also removing StatsBase from the notebook environment, and so Pluto rightly asks you to restart the notebook to sync things up. I think the act of the notebook updating after a cell is deleted can been seen a bit more clearly here:

https://user-images.githubusercontent.com/25312320/155642367-beb4f270-2204-4bb5-9ec4-a5b505532e75.mp4

So to get the expected behavior when moving packages around, I would just do all my moving first, and then run all the cells at once by pressing Ctrl+S on the keyboard after I am finished:

https://user-images.githubusercontent.com/25312320/155642171-267ad568-d0f7-4905-abc6-411bfa997a1b.mp4

btw, thanks for including a video. It made it a lot easier to see what you were trying to do!

**Reply by yha** [reply]
Then it seems like the problem is more general: I'm surprised Pluto re-runs things at all when deleting a cell. The first video looks to me like a bug as well, although a less annoying one.
Is there a reason why deleting a cell should trigger running code?

**Reply by yha** [reply]
Actually, thinking about it some more, I don't understand how the first video relates to the original issue. 
In the original video, when the cell is deleted, the list of packages is exactly the same as it was on the last save. So the notebook syncing doesn't explain why packages are resolved.

**Reply by icweaver** [reply]
I could be misunderstanding what you mean here, but the re-running of dependent cells after a change has been made (including deletion) seems like a feature or reactivity to me, rather than a bug. Using the first video I recorded as an example, wouldn't the state of the notebook become hidden if `x` remained defined after deleting the cell that created it in the first place?

**Reply by icweaver** [reply]
> In the original video, when the cell is deleted, the list of packages is exactly the same as it was on the last save

I don't think this is true, doing a quick test seems to suggest otherwise:

https://user-images.githubusercontent.com/25312320/155786106-22e719db-9484-4867-9bf3-cc2d9e234863.mp4

I think this can also be seen more clearly using Pluto's "secret" `auto_reload_from_file` feature:

https://user-images.githubusercontent.com/25312320/155797476-0a7cc26e-8058-4283-af36-2d2f31de08fd.mp4

But just to make sure I am sharing the same terminology as you, by "cell deletion" here I mean when the empty cell is actually run again (either by running the empty cell explicitly or hitting backspace inside of it to delete the cell entirely)

**Reply by yha** [reply]
> I could be misunderstanding what you mean here, but the re-running of dependent cells after a change has been made (including deletion) seems like a feature or reactivity to me, rather than a bug. Using the first video I recorded as an example, wouldn't the state of the notebook become hidden if `x` remained defined after deleting the cell that created it in the first place?

But Pluto doesn't run dependent cells after any change -- that would be super annoying. For example, it doesn't rerun code after each character typed, or even on typing a newline or moving to a different cell. So I think it's surprising that hitting backspace to delete a cell triggers running code. In fact, since the cell is already empty, I expect that deleting it would be merely a cosmetic change.

> I don't think this is true, doing a quick test seems to suggest otherwise:

I don't understand what this test is supposed to show. Sure, the list of packages changes in this example, but not in the original example that I posted as an issue. In my original video, the notebook has been saved with three `using` statements: `Unzip`, `Statistics` and `StatsBase`. When the cell is deleted, the code still contains these same three `using` statements.

If I understand correctly, the problem is that deleting the cell causes Pluto to update dependent cells and then the package list, and the first cell happens not be a dependent cell so there's now an out-of-date package list based on a partial update. If that is correct, I think that's a strong argument against running code in response to deleting a cell.

**Reply by icweaver** [reply]
> I don't understand what this test is supposed to show. Sure, the list of packages changes in this example, but not in the original example that I posted as an issue.

It does though:

https://user-images.githubusercontent.com/25312320/155799318-ce628fdf-16fb-42e4-8788-24172e980b4a.mp4

Deleting the second cell removes `StatsBase` from your notebook, and you never ran the remaining cell with all three packages present, so the notebook frontend and file are now out of sync, which is why a restart is required

**Reply by icweaver** [reply]
> But Pluto doesn't run dependent cells after any change -- that would be super annoying. For example, it doesn't rerun code after each character typed, or even on typing a newline or moving to a different cell

I think we are using different definitions of "change" here. By change, I mean when the notebook is saved. Like you said, this does not happen while you are typing. Only when the cell is explicitly run

> So I think it's surprising that hitting backspace to delete a cell triggers running code. In fact, since the cell is already empty, I expect that deleting it would be merely a cosmetic change.

This is not a cosmetic change. You are actually changing the contents of the notebook (deleting the cell on file) when you do this, as shown above

**Reply by yha** [reply]
> You are actually changing the contents of the notebook (deleting the cell on file) when you do this, as shown above

I realize that now, and this is what I have issue with. I think that deleting a cell should not trigger a change.
Your initial suggestion was "I would just do all my moving first, and then run all the cells at once by pressing ctrl+S on the keyboard after I am finished". In my mind, that was exactly what I was doing. I expected to do my changes, including cell deletions, and press ctrl+S when I'm finished. Pluto surprised me by responding to the cell deletion with a partial update (of that cell only).
This is especially surprising since a cell can be deleted by simply hitting backspace. No other text editing commands trigger a change.

**Reply by icweaver** [reply]
> Your initial suggestion was "I would just do all my moving first, and then run all the cells at once by pressing ctrl+S on the keyboard after I am finished".

To be clear, I think this was my one and only suggestion

**Reply by icweaver** [reply]
I can totally relate to being surprised by some of the things that Pluto does, especially when trying to make it fit into a "Jupyter/traditional text editor" mental model, but I don't think that is necessarily a very productive way to approach novel things. I hope that the examples above helped make things a little less surprising, but I am more than happy to consider alternatives with you. Taking the last mini example I showed as a concrete example, how would you like the notebook file to behave after the second cell has been deleted (after backspace was hit inside of its empty cell)?

**Reply by yha** [reply]
> I hope that the examples above helped make things a little less surprising

Definitely. I understand now how this behavior comes about.

> Taking the last mini example I showed as a concrete example, how would you like the notebook file to behave after the second cell has been deleted (after backspace was hit inside of its empty cell)?

I would expect to see the cell deletion in the frontend, and that's it: no triggering of code or changes to the notebook file, until I explicitly run code (ctrl/shift+enter) or press ctrl+S.

**Reply by icweaver** [reply]
For sure. Ok, happy to try this to see what we can shake out. So after deleting the cell containing `StatsBase` and then pasting it into the first cell, you would like to have something like this right, where nothing in the file has been updated yet and the first cell has not explicitly been re-run? (Sorry for the potato quality, I'm just screenshotting from the video above):

![Screenshot from 2022-02-27 15-47-01](https://user-images.githubusercontent.com/25312320/155899298-83efc170-ae88-429b-b1ad-30083dd43e17.png)

Moving forward, if a new cell were to then be added again, say with an import of a completely different package, and then all of the cells explicitly run with Ctrl-S, how would we handle this update? It's not completely clear to me how we could accomplish this cleanly since now our front-end and notebook file are currently out of sync.

**Reply by icweaver** [reply]
Actually, as a simpler scenario, if we just started from here:

![Screenshot from 2022-02-27 16-29-47](https://user-images.githubusercontent.com/25312320/155900808-94afd1da-5eac-4943-aa98-bf9e825df9a3.png)

removed the `StatsBase` cell, and then closed the notebook, which state should be saved?

**Reply by yha** [reply]
The state before removing the cell. Deleting the cell should behave the same as deleting all the code inside the cell (and there should be a browser prompt when closing the notebook -- the same one you already get when closing a notebook with unsaved changes)

**Reply by icweaver** [reply]
Ok, so you are suggesting to introduce hidden state. I could be mistaken, but I think this goes against one of the core philosophies of Pluto that what you see is what you get (i.e., the state of the notebook is completely expressed by what you see when you are viewing it). It seems like the issue here just comes down to a difference of opinion, which is totally fine.

I hope that the (in my opinion) [small adjustment](https://github.com/fonsp/Pluto.jl/discussions/1955#discussioncomment-2253156) to how to approach cell deletion in Pluto that I shared earlier can be of use, but I understand if you are looking for a different project with a different design philosophy. At any rate, thanks for the good discussion, I think having these types of conversations are really valuable for considering new ideas!

**Reply by yha** [reply]
> Ok, so you are suggesting to introduce hidden state.

No, I'm not. Pluto already has hidden state: when you edit code, the actual state of the notebook file is hidden, until the notebook is explicitly saved and brought to sync. To have no hidden state, Pluto would need to save the notebook after each keystroke. 
I'm just suggesting that deleting a cell should behave like editing code. I fail to see why they should be different.

**Reply by icweaver** [reply]
> I fail to see why they should be different

Again, I think the disconnect is coming from what we consider [to be a change](https://github.com/fonsp/Pluto.jl/discussions/1955#discussioncomment-2253925). Under this definition, I mean that the notebook state is the same _once a cell is run_, and thus [not hidden](https://github.com/fonsp/Pluto.jl#dynamic-environment). We just seem to have a difference of opinion of when this change should be registered. I don't think there is much more that I can say on this matter, but I really do wish you the best regarding this issue

**Reply by yha** [reply]
I'm sorry to keep insisting, but I still don't understand your position.
Can you explain why you think deleting a line of code should be different to deleting a cell?

**Reply by icweaver** [reply]
Ok, just to make sure we are on the same page here, I am assuming that by "editing code" you mean before hitting save/running the cell.

I see deleting a cell as different from this case because you are actually removing information from the dependency graph, which needs to be accounted for in some way. Just using the same video  I shared above again as an example for convenience:

https://user-images.githubusercontent.com/25312320/156058423-f83fd336-0c19-4880-9483-16217b376079.mp4

Before removing the empty cell, no changes have been committed (the cell is highlighted in yellow as a visual cue that once you run this cell, the current state will be altered), and so the dependency graph is unchanged. So far everything is the same between deleting a line of code and "deleting a cell but not committing that change just yet". Once you go the extra step of actually deleting the cell, we have fully committed to altering the dependency graph. That is the difference.

Now whether you agree with this behavior or not is totally up to you, but as it stands, I believe this is fully in line with the reactive, reproducible, and non-hidden state that Pluto notebooks promise.

**Reply by yha** [reply]
> So far everything is the same between deleting a line of code and "deleting a cell but not committing that change just yet".

I think we are still using different terminology here. When I refer to "deleting a cell", I do not mean deleting its contents. I mean pressing backspace in an empty cell or selecting "Delete cell" from the cell's "..." menu. So currently there is no such thing in Pluto as "deleting a cell but not committing that change just yet", and I think there should be.

> Once you go the extra step of actually deleting the cell, we have fully committed to altering the dependency graph. 

This is the behavior of Pluto I'm trying to understand the rationale for, and you seem to be using it as an assumption.

**Reply by icweaver** [reply]
> I think we are still using different terminology here. When I refer to "deleting a cell", I do not mean deleting its contents. I mean pressing backspace in an empty cell or selecting "Delete cell" from the cell's "..." menu.

I think we are on the same page about this.

> So currently there is no such thing in Pluto as "deleting a cell but not committing that change just yet"

Yes, that's why I used quotations, but I could have made that more clear

> and I think there should be

This is where we diverge

> This is the behavior of Pluto I'm trying to understand the rationale for, and you seem to be using it as an assumption.

I think it would be unreasonable to expect folks to be immediately comfortable with the quirks of Pluto, especially when coming from editors that follow a more traditional paradigm, so I appreciate you taking the time to try and approach things from a different perspective =]

**Reply by yha** [reply]
> I think it would be unreasonable to expect folks to be immediately comfortable with the quirks of Pluto, especially when coming from editors that follow a more traditional paradigm

Perhaps I should clarify that I am not a newcomer to Pluto. For the last half year or so, I've been using it almost every workday, and this still tripped me up.
I still don't see any downside to my proposed change (apart from the work involved in implementing it, obviously), but I'm willing to let it go at this point, and just get used to being careful about deleting cells when there are unsaved changes. Thank you very much for your time.

**Reply by icweaver** [reply]
I'm sorry, I never meant to imply that you were. I think different aspects of Pluto can trip up users of any experience level, and that there can always be room for improvement in this regard, so thank you for sharing your experience. I have been discussing how to handle this moving forward with Fons, and share some potential next steps below:

> Fons van der Plas: anyways! their suggestion does make sense: allow you to delete a cell without committing, just like you can currently delete a line within a cell without committing. This strategy of "making multiple edits, and running them later with a single Ctrl+S" is actually really useful, so that would be cool.
> 
> Fons van der Plas: But! I think it would be confusing without new UI, we would clearly need to show "a cell was deleted here, you still need to save it". And this might be perceived as annoying ("why doesn't Pluto just save it?")
> 
> Fons van der Plas: We could also go at the underlying issue (moving imports around can lead to "restart required") in another way, by making pluto keep track of which imports at which versions existed before. Perhaps this could be done by Julia itself, there is some ongoing work in this direction: https://github.com/JuliaLang/Pkg.jl/pull/3002 (read "sysimage" as "already loaded in the process") and https://github.com/JuliaLang/julia/pull/44329

It might be better to move this to a separate thread or the Zulip, but will leave this up to you!

**Comment by fonsp** [comment]
Hey all! Very interesting question! I have seen other people confused/annoyed by this behaviour.

Like pointed out in the discussion above (I only read parts of it): the behaviour is "intentional": our current model is that deleting a cell should immediately "commit" that change, hence causing this issue. 

---

The feature suggestion by @yha makes a lot of sense: allow you to delete a cell without committing, just like you can currently delete a line within a cell without committing. This strategy of "making multiple edits, and running them later with a single Ctrl+S" is actually really useful, so that would be cool.

But! It would need some good new UI to pull this off, and I imagine that this new behaviour would then become a source of confusion/annoyance for its own reasons. (Especially for new users.)

---

We could also go at the underlying issue (moving imports around can lead to "restart required") in another way, by keeping track of which imports at which versions existed before. That way, removing an import, committing, and then adding it, committing, would become a no-op as a whole. This "restart recommended" message would appear, and then disappear again.

But! I think this would be too difficult to implement correctly, especially because we try to only expose public Pkg API through GUI, without touching internals. (And the gain might be too small.) Now, if this were done by Julia itself, then that might be the way to go! There is some ongoing work in this direction: https://github.com/JuliaLang/Pkg.jl/pull/3002 (read "sysimage" as "already loaded in the process") and https://github.com/JuliaLang/julia/pull/44329 . 

---

All in all, I would say that right now, this is something we have to live with, because a fix would be too difficult and time-consuming to implement. But I can imagine a future where we have more API to work with from Pkg and Julia.

Specifically: 
- API like `which_packages_are_loaded_with_versions()` would make it much easier to check whether a restart is recommended or required, by simply comparing this list with the Manifest.toml. Right now we have to jump through lots of heuristic hoops to get it.
- API like `Pkg.add(name; only_allow_versions_already_loaded=true)` would make sure that removing, and then re-adding a package will not change its version in the manifest.
